feat(HTTP): Refresh 3XX response pages (#35353)
mdn/content: files/en-us/web/http/status/304/index.md

{{HTTPSidebar}}

> The HTTP **`304 Not Modified`** [redirection response](/en-US/docs/Web/HTTP/Status#redirection_messages) status code indicates that there is no need to retransmit the requested resources.
> This response code is sent when the request is a [conditional](/en-US/docs/Web/HTTP/Conditional_requests) {{HTTPMethod("GET")}} or {{HTTPMethod("HEAD")}} request with an {{HTTPHeader("If-None-Match")}} or an {{HTTPHeader("If-Modified-Since")}} header and the condition evaluates to 'false'.
> It confirms that the resource cached by the client is still valid and that the server would have sent a {{HTTPStatus("200", "200 OK")}} response with the resource if the condition evaluated to 'true'.

Reviewer: FWIW, I like to clearly separate a summary sentence "what is it for" and the explanation. This is "to my taste" only. IMO it makes it easier for readers to parse and decide relevance.

```suggestion
The HTTP **`304 Not Modified`** [redirection response](/en-US/docs/Web/HTTP/Status#redirection_messages) status code indicates that there is no need to retransmit the requested resources.

This response code is sent when the request is a [conditional](/en-US/docs/Web/HTTP/Conditional_requests) {{HTTPMethod("GET")}} or {{HTTPMethod("HEAD")}} request with an {{HTTPHeader("If-None-Match")}} or an {{HTTPHeader("If-Modified-Since")}} header and the condition evaluates to 'false'.
It confirms that the resource cached by the client is still valid and that the server would have sent a {{HTTPStatus("200", "200 OK")}} response with the resource if the condition evaluated to 'true'.
```

I will try not to mention this again :-)

Author: I'm fine with this!

## Examples

### 304 response to conditional requests

> The examples below show {{HTTPMethod("GET")}} requests made using [curl](https://curl.se/) with conditional request headers and the HTTP responses received in return.
> The first example would return a {{HTTPStatus("200", "200 OK")}} if we know the resource has been updated since the timestamp in the {{HTTPHeader("If-Modified-Since")}} header.

Reviewer: ```suggestion
The examples below show {{HTTPMethod("GET")}} requests made using [curl](https://curl.se/) with conditional request headers and the HTTP responses received in return.

The first example would return a {{HTTPStatus("200", "200 OK")}} if we know the resource has been updated since the timestamp in the {{HTTPHeader("If-Modified-Since")}} header.
```

## Examples

### 304 response to conditional requests

The examples below show {{HTTPMethod("GET")}} requests made using [curl](https://curl.se/) with conditional request headers.
The `--http1.1` flag is used to force the HTTP/1.1 protocol for readability.

> The request uses a future date of 21st November 2050 to check whether if the resource has been updated after this date:

Reviewer: ```suggestion

The first request uses an `If-Modified-Since` condition with a future date of 21st November 2050.
This must evaluate to `false`, because the resource can't have been updated after a time that hasn't happened yet:
```

```bash
curl --http1.1 -I --header 'If-Modified-Since: Tue, 21 Nov 2050 08:00:00 GMT' \
 https://developer.mozilla.org/en-US/
```

> This will perform the following request:

Reviewer: ```suggestion
This will result in the following HTTP request:
```

```http
GET /en-US/ HTTP/1.1
Host: developer.mozilla.org
User-Agent: curl/8.7.1
Accept: */*
If-Modified-Since: Tue, 21 Nov 2050 08:00:00 GMT
```

The response would be {{HTTPStatus("200", "200 OK")}} if the resource was updated after the timestamp in the {{HTTPHeader("If-Modified-Since")}} header.

> Instead, we have a `304` response that includes {{HTTPHeader("ETag")}}, {{HTTPHeader("Age")}} and {{HTTPHeader("Expires")}} headers:

Reviewer: ```suggestion
Instead, we get a `304` response that includes {{HTTPHeader("ETag")}}, {{HTTPHeader("Age")}} and {{HTTPHeader("Expires")}} headers:
```

> The response would be {{HTTPStatus("200", "200 OK")}} if the resource was updated after the timestamp in the {{HTTPHeader("If-Modified-Since")}} header.
> Instead, we have a `304` response that includes {{HTTPHeader("ETag")}}, {{HTTPHeader("Age")}} and {{HTTPHeader("Expires")}} headers:

Reviewer: ```suggestion
The response would be {{HTTPStatus("200", "200 OK")}} with the current version of the resource if the resource had been updated after the timestamp in the {{HTTPHeader("If-Modified-Since")}} header.
Instead, we get a `304` response that includes {{HTTPHeader("ETag")}}, {{HTTPHeader("Age")}} and {{HTTPHeader("Expires")}} headers, telling us our cached version of the resource is still current:
```

```http
HTTP/1.1 304 Not Modified
Date: Wed, 28 Aug 2024 09:52:35 GMT
Expires: Wed, 28 Aug 2024 10:01:53 GMT
Age: 3279
ETag: "b20a0973b226eeea30362acb81f9e0b3"
Cache-Control: public, max-age=3600
Vary: Accept-Encoding
X-cache: hit
Alt-Svc: clear
```

> If you run another `curl` command with a fresh `etag` value from the previous response in an {{HTTPHeader("If-None-Match")}} header, you receive a `304 Not Modified` if the request contains :
> 
> ```bash

Reviewer: This is a little confusing - in particular what does "fresh" mean?

```suggestion
Now run another `curl` command using the `etag` value from the previous response with the {{HTTPHeader("If-None-Match")}} condition (since this `etag` is the current version of the resource on the server we expect to receive a `304 Not Modified` response):
```

> curl --http1.1 -I --header 'If-None-Match: "e27d81b845c3716cdb5d4220d78e2799"' \

Reviewer: To be consistent, this has to be:

```suggestion
curl --http1.1 -I --header 'If-None-Match: "b20a0973b226eeea30362acb81f9e0b3"' \
```

 https://developer.mozilla.org/en-US/
```

> This will perform the following request:

Reviewer: ```suggestion
This will result in the following HTTP  request:
```

expires: Tue, 21 Nov 2023 08:53:14 GMT
age: 3194
etag: "e27d81b845c3716cdb5d4220d78e2799"
cache-control: public, max-age=3600
```

A `304 Not Modified` response is also returned in response to a {{HTTPMethod("GET")}} request containing an {{HTTPHeader("If-None-Match")}} header with the {{HTTPHeader("ETag")}} from the response above.
Because the `etag` value exists, a matching entity tag fails the condition, and a `304` response is returned:

```bash
curl -I --header 'If-None-Match: "e27d81b845c3716cdb5d4220d78e2799"' \
 https://developer.mozilla.org/en-US/
```

> 

Reviewer: Also, consider making these separate examples with headings? I don't like the syntax `> Request` in the code block below because it looks like part of the HTTP message. Ditto for the response. I can see why it has been done, but I think it is not good.

Author: true, I think the output is formatted that way in curl, I've reworked in 2771552629bf8ba63be74b5f5830ea93982707ef

Author: Also adding `--http1.1` flag to curl
